File: sdk/csharp/tests/AnyVali.Tests/InterchangeTests.cs
using AnyVali;
using AnyVali.Schemas;
using Xunit;

namespace AnyVali.Tests;

public class ExportTests
{
    [Fact]
    public void ExportStringSchema()
    {
        var s = V.String().MinLength(1).MaxLength(100);
        var doc = s.Export();
        Assert.Equal("1.0", doc.AnyvaliVersion);
        Assert.Equal("1", doc.SchemaVersion);
        Assert.Equal("string", doc.Root["kind"]);
        Assert.Equal(1L, doc.Root["minLength"]);
        Assert.Equal(100L, doc.Root["maxLength"]);
    }

    [Fact]
    public void ExportNumberSchema()
    {
        var s = V.Number().Min(0).Max(100);
        var doc = s.Export();
        Assert.Equal("number", doc.Root["kind"]);
        Assert.Equal(0.0, doc.Root["min"]);
        Assert.Equal(100.0, doc.Root["max"]);
    }

    [Fact]
    public void ExportIntSchema()
    {
        var s = V.Int().Min(0);
        var doc = s.Export();
        Assert.Equal("int", doc.Root["kind"]);
    }

    [Fact]
    public void ExportObjectSchema()
    {
        var s = V.Object(new Dictionary<string, Schema>
        {
            ["name"] = V.String(),
            ["age"] = V.Optional(V.Int()),
        });
        var doc = s.Export();
        Assert.Equal("object", doc.Root["kind"]);
        var required = (List<object?>)doc.Root["required"]!;
        Assert.Contains("name", required);
        Assert.DoesNotContain("age", required);
    }

    [Fact]
    public void ExportArraySchema()
    {
        var s = V.Array(V.String()).MinItems(1);
        var doc = s.Export();
        Assert.Equal("array", doc.Root["kind"]);
        Assert.Equal(1L, doc.Root["minItems"]);
    }

    [Fact]
    public void ExportUnionSchema()
    {
        var s = V.Union(V.String(), V.Int());
        var doc = s.Export();
        Assert.Equal("union", doc.Root["kind"]);
    }

    [Fact]
    public void ExportWithDefault()
    {
        var s = V.String().Default("hello");
        var doc = s.Export();
        Assert.Equal("hello", doc.Root["default"]);
    }

    [Fact]
    public void PortableExportFailsForNonPortable()
    {
        var s = V.String();
        s.IsPortable = false;
        Assert.Throws<InvalidOperationException>(() => s.Export(ExportMode.Portable));
    }

    [Fact]
    public void ExtendedExportAllowed()
    {
        var s = V.String();
        s.IsPortable = false;
        var doc = s.Export(ExportMode.Extended);
        Assert.NotNull(doc);
    }

    [Fact]
    public void ExportDocumentToJson()
    {
        var s = V.String();
        var doc = s.Export();
        var json = doc.ToJson();
        Assert.Contains("\"kind\"", json);
        Assert.Contains("\"string\"", json);
    }
}

public class ImportTests
{
    [Fact]
    public void ImportStringSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "string", ["minLength"] = 1L },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal("hello", schema.Parse("hello"));
        var result = schema.SafeParse("");
        Assert.False(result.Success);
    }

    [Fact]
    public void ImportNumberSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "number", ["min"] = 0.0 },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal(5.0, schema.Parse(5L));
        var result = schema.SafeParse(-1L);
        Assert.False(result.Success);
    }

    [Fact]
    public void ImportIntSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "int", ["min"] = 1L },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal(5L, schema.Parse(5L));
    }

    [Fact]
    public void ImportObjectSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?>
            {
                ["kind"] = "object",
                ["properties"] = new Dictionary<string, object?>
                {
                    ["name"] = new Dictionary<string, object?> { ["kind"] = "string" },
                    ["age"] = new Dictionary<string, object?> { ["kind"] = "int" },
                },
                ["required"] = new List<object?> { "name", "age" },
                ["unknownKeys"] = "reject",
            },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        var input = new Dictionary<string, object?> { ["name"] = "Alice", ["age"] = 30L };
        var result = (Dictionary<string, object?>)schema.Parse(input)!;
        Assert.Equal("Alice", result["name"]);
    }

    [Fact]
    public void ImportArraySchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?>
            {
                ["kind"] = "array",
                ["items"] = new Dictionary<string, object?> { ["kind"] = "string" },
            },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        var input = new List<object?> { "a", "b" };
        var result = (List<object?>)schema.Parse(input)!;
        Assert.Equal(2, result.Count);
    }

    [Fact]
    public void ImportUnionSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?>
            {
                ["kind"] = "union",
                ["variants"] = new List<object?>
                {
                    new Dictionary<string, object?> { ["kind"] = "string" },
                    new Dictionary<string, object?> { ["kind"] = "int" },
                },
            },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal("hello", schema.Parse("hello"));
        Assert.Equal(42L, schema.Parse(42L));
    }

    [Fact]
    public void ImportWithDefault()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "string", ["default"] = "fallback" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        var result = schema.SafeParse(AnyVali.Parse.Absent.Value);
        Assert.True(result.Success);
        Assert.Equal("fallback", result.Data);
    }

    [Fact]
    public void ImportWithCoercion()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "int", ["coerce"] = "string->int" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal(42L, schema.Parse("42"));
    }

    [Fact]
    public void ImportRefSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "ref", ["ref"] = "#/definitions/MyStr" },
            Definitions = new Dictionary<string, object?>
            {
                ["MyStr"] = new Dictionary<string, object?> { ["kind"] = "string" },
            },
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal("hello", schema.Parse("hello"));
    }

    [Fact]
    public void ImportNullableSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?>
            {
                ["kind"] = "nullable",
                ["schema"] = new Dictionary<string, object?> { ["kind"] = "string" },
            },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Null(schema.Parse(null));
        Assert.Equal("hello", schema.Parse("hello"));
    }

    [Fact]
    public void ImportOptionalSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?>
            {
                ["kind"] = "optional",
                ["schema"] = new Dictionary<string, object?> { ["kind"] = "string" },
            },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        var result = schema.SafeParse(AnyVali.Parse.Absent.Value);
        Assert.True(result.Success);
    }

    [Fact]
    public void ImportTupleSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?>
            {
                ["kind"] = "tuple",
                ["elements"] = new List<object?>
                {
                    new Dictionary<string, object?> { ["kind"] = "string" },
                    new Dictionary<string, object?> { ["kind"] = "int" },
                },
            },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        var input = new List<object?> { "hello", 42L };
        var result = (List<object?>)schema.Parse(input)!;
        Assert.Equal("hello", result[0]);
        Assert.Equal(42L, result[1]);
    }

    [Fact]
    public void ImportRecordSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?>
            {
                ["kind"] = "record",
                ["values"] = new Dictionary<string, object?> { ["kind"] = "int" },
            },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        var input = new Dictionary<string, object?> { ["a"] = 1L, ["b"] = 2L };
        var result = (Dictionary<string, object?>)schema.Parse(input)!;
        Assert.Equal(1L, result["a"]);
    }

    [Fact]
    public void ImportIntersectionSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?>
            {
                ["kind"] = "intersection",
                ["allOf"] = new List<object?>
                {
                    new Dictionary<string, object?>
                    {
                        ["kind"] = "object",
                        ["properties"] = new Dictionary<string, object?>
                        {
                            ["a"] = new Dictionary<string, object?> { ["kind"] = "string" },
                        },
                        ["required"] = new List<object?> { "a" },
                        ["unknownKeys"] = "strip",
                    },
                    new Dictionary<string, object?>
                    {
                        ["kind"] = "object",
                        ["properties"] = new Dictionary<string, object?>
                        {
                            ["b"] = new Dictionary<string, object?> { ["kind"] = "int" },
                        },
                        ["required"] = new List<object?> { "b" },
                        ["unknownKeys"] = "strip",
                    },
                },
            },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        var input = new Dictionary<string, object?> { ["a"] = "hello", ["b"] = 42L };
        var result = (Dictionary<string, object?>)schema.Parse(input)!;
        Assert.Equal("hello", result["a"]);
        Assert.Equal(42L, result["b"]);
    }

    [Fact]
    public void ImportLiteralSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "literal", ["value"] = "hello" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal("hello", schema.Parse("hello"));
    }

    [Fact]
    public void ImportEnumSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?>
            {
                ["kind"] = "enum",
                ["values"] = new List<object?> { "a", "b", "c" },
            },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal("b", schema.Parse("b"));
    }

    [Fact]
    public void ImportBoolSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "bool" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal(true, schema.Parse(true));
    }

    [Fact]
    public void ImportNullSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "null" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Null(schema.Parse(null));
    }

    [Fact]
    public void ImportAnySchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "any" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal("anything", schema.Parse("anything"));
    }

    [Fact]
    public void ImportNeverSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "never" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        var result = schema.SafeParse("anything");
        Assert.False(result.Success);
    }

    [Fact]
    public void ImportUnsupportedKindThrows()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "not_a_real_kind" },
            Definitions = new(),
            Extensions = new(),
        };
        Assert.Throws<InvalidOperationException>(() => V.Import(doc));
    }

    [Fact]
    public void RoundTripExportImport()
    {
        var original = V.Object(new Dictionary<string, Schema>
        {
            ["name"] = V.String().MinLength(1),
            ["age"] = V.Optional(V.Int().Min(0)),
        });

        var doc = original.Export();
        var imported = V.Import(doc);

        var input = new Dictionary<string, object?> { ["name"] = "Alice", ["age"] = 30L };
        var result = imported.SafeParse(input);
        Assert.True(result.Success);
    }

    [Fact]
    public void ImportAllIntTypes()
    {
        foreach (var kind in new[] { "int8", "int16", "int32", "int64", "uint8", "uint16", "uint32", "uint64" })
        {
            var doc = new AnyValiDocument
            {
                AnyvaliVersion = "1.0",
                SchemaVersion = "1",
                Root = new Dictionary<string, object?> { ["kind"] = kind },
                Definitions = new(),
                Extensions = new(),
            };
            var schema = V.Import(doc);
            Assert.Equal(0L, schema.Parse(0L));
        }
    }

    [Fact]
    public void ImportFloat32Schema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "float32" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal(1.5, schema.Parse(1.5));
    }

    [Fact]
    public void ImportFloat64Schema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "float64" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal(1.5, schema.Parse(1.5));
    }

    [Fact]
    public void ImportUnknownSchema()
    {
        var doc = new AnyValiDocument
        {
            AnyvaliVersion = "1.0",
            SchemaVersion = "1",
            Root = new Dictionary<string, object?> { ["kind"] = "unknown" },
            Definitions = new(),
            Extensions = new(),
        };
        var schema = V.Import(doc);
        Assert.Equal("x", schema.Parse("x"));
    }
}

File: sdk/rust/tests/interchange_tests.rs
use anyvali::*;
use serde_json::json;
use std::collections::HashMap;

#[test]
fn export_string_schema() {
    let s = string().min_length(1).max_length(100);
    let doc = export(&s, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.anyvali_version, "1.0");
    assert_eq!(doc.schema_version, "1");
    assert_eq!(doc.root["kind"], "string");
    assert_eq!(doc.root["minLength"], 1);
    assert_eq!(doc.root["maxLength"], 100);
}

#[test]
fn export_number_schema() {
    let n = number().min(0.0).max(100.0);
    let doc = export(&n, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "number");
}

#[test]
fn export_int_schema() {
    let i = int().min(1.0).max(10.0);
    let doc = export(&i, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "int");
}

#[test]
fn export_object_schema() {
    let o = object()
        .field("name", Box::new(string()))
        .field("age", Box::new(int()))
        .required(vec!["name", "age"]);
    let doc = export(&o, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "object");
    assert!(doc.root["properties"]["name"].is_object());
    assert!(doc.root["properties"]["age"].is_object());
}

#[test]
fn export_array_schema() {
    let a = array(Box::new(string())).min_items(1);
    let doc = export(&a, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "array");
    assert_eq!(doc.root["items"]["kind"], "string");
    assert_eq!(doc.root["minItems"], 1);
}

#[test]
fn export_with_definitions() {
    let s = ref_("#/definitions/User");
    let mut defs: HashMap<String, Box<dyn Schema>> = HashMap::new();
    defs.insert(
        "User".to_string(),
        Box::new(
            object()
                .field("name", Box::new(string()))
                .required(vec!["name"]),
        ),
    );
    let doc = export(&s, ExportMode::Portable, &defs).unwrap();
    assert!(doc.definitions.contains_key("User"));
    assert_eq!(doc.definitions["User"]["kind"], "object");
}

#[test]
fn export_union() {
    let u = union(vec![Box::new(string()), Box::new(int())]);
    let doc = export(&u, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "union");
    assert_eq!(doc.root["variants"].as_array().unwrap().len(), 2);
}

#[test]
fn export_intersection() {
    let i = intersection(vec![Box::new(number().min(0.0)), Box::new(number().max(100.0))]);
    let doc = export(&i, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "intersection");
}

#[test]
fn export_nullable() {
    let n = nullable(Box::new(string()));
    let doc = export(&n, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "nullable");
    assert_eq!(doc.root["schema"]["kind"], "string");
}

#[test]
fn export_optional() {
    let o = optional(Box::new(int()));
    let doc = export(&o, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "optional");
}

#[test]
fn export_literal() {
    let l = literal(json!("hello"));
    let doc = export(&l, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "literal");
    assert_eq!(doc.root["value"], "hello");
}

#[test]
fn export_enum() {
    let e = enum_(vec![json!("a"), json!("b")]);
    let doc = export(&e, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "enum");
    assert_eq!(doc.root["values"], json!(["a", "b"]));
}

#[test]
fn export_tuple() {
    let t = tuple(vec![Box::new(string()), Box::new(int())]);
    let doc = export(&t, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "tuple");
    assert_eq!(doc.root["elements"].as_array().unwrap().len(), 2);
}

#[test]
fn export_record() {
    let r = record(Box::new(string()));
    let doc = export(&r, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "record");
    assert_eq!(doc.root["values"]["kind"], "string");
}

#[test]
fn export_ref() {
    let r = ref_("#/definitions/User");
    let doc = export(&r, ExportMode::Portable, &HashMap::new()).unwrap();
    assert_eq!(doc.root["kind"], "ref");
    assert_eq!(doc.root["ref"], "#/definitions/User");
}

#[test]
fn export_primitives() {
    for (schema, expected_kind): (Box<dyn Schema>, &str) in vec![
        (Box::new(any()) as Box<dyn Schema>, "any"),
        (Box::new(unknown()), "unknown"),
        (Box::new(never()), "never"),
        (Box::new(null()), "null"),
        (Box::new(bool_()), "bool"),
    ] {
        let doc = export(schema.as_ref(), ExportMode::Portable, &HashMap::new()).unwrap();
        assert_eq!(doc.root["kind"], expected_kind);
    }
}

// ===== Import Tests =====

#[test]
fn import_string_schema() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "string", "minLength": 3 },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert_eq!(schema.kind(), "string");
    assert!(schema.parse(&json!("abc")).is_ok());
    assert!(!schema.safe_parse(&json!("ab")).success);
}

#[test]
fn import_number_schema() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "number", "min": 0, "max": 100 },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!(50)).is_ok());
    assert!(!schema.safe_parse(&json!(-1)).success);
}

#[test]
fn import_int_schema() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "int" },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!(42)).is_ok());
    assert!(!schema.safe_parse(&json!(3.14)).success);
}

#[test]
fn import_object_schema() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": {
            "kind": "object",
            "properties": {
                "name": { "kind": "string" },
                "age": { "kind": "int" }
            },
            "required": ["name", "age"],
            "unknownKeys": "reject"
        },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!({"name": "Alice", "age": 30})).is_ok());
}

#[test]
fn import_with_definitions() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": {
            "kind": "object",
            "properties": {
                "user": { "kind": "ref", "ref": "#/definitions/User" }
            },
            "required": ["user"],
            "unknownKeys": "reject"
        },
        "definitions": {
            "User": {
                "kind": "object",
                "properties": {
                    "name": { "kind": "string" }
                },
                "required": ["name"],
                "unknownKeys": "reject"
            }
        },
        "extensions": {}
    }"#;
    let (schema, ctx) = import(json_str).unwrap();
    let result = schema.parse_with_context(
        &json!({"user": {"name": "Alice"}}),
        &ctx,
    );
    assert!(result.is_ok());
}

#[test]
fn import_coerce_config() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "int", "coerce": "string->int" },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    let result = schema.parse(&json!("42"));
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), json!(42));
}

#[test]
fn import_coerce_array_config() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "string", "coerce": ["trim", "lower"] },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    let result = schema.parse(&json!("  HELLO  "));
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), json!("hello"));
}

#[test]
fn import_all_int_widths() {
    for kind in &["int8", "int16", "int32", "int64", "uint8", "uint16", "uint32", "uint64"] {
        let json_str = format!(
            r#"{{"anyvaliVersion":"1.0","schemaVersion":"1","root":{{"kind":"{}"}},"definitions":{{}},"extensions":{{}}}}"#,
            kind
        );
        let (schema, _ctx) = import(&json_str).unwrap();
        assert_eq!(schema.kind(), *kind);
    }
}

#[test]
fn import_float_widths() {
    for kind in &["float32", "float64"] {
        let json_str = format!(
            r#"{{"anyvaliVersion":"1.0","schemaVersion":"1","root":{{"kind":"{}"}},"definitions":{{}},"extensions":{{}}}}"#,
            kind
        );
        let (schema, _ctx) = import(&json_str).unwrap();
        assert_eq!(schema.kind(), *kind);
    }
}

#[test]
fn import_all_simple_kinds() {
    for kind in &["any", "unknown", "never", "null", "bool"] {
        let json_str = format!(
            r#"{{"anyvaliVersion":"1.0","schemaVersion":"1","root":{{"kind":"{}"}},"definitions":{{}},"extensions":{{}}}}"#,
            kind
        );
        let (schema, _ctx) = import(&json_str).unwrap();
        assert_eq!(schema.kind(), *kind);
    }
}

#[test]
fn import_literal() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "literal", "value": "hello" },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!("hello")).is_ok());
    assert!(!schema.safe_parse(&json!("world")).success);
}

#[test]
fn import_enum() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "enum", "values": ["a", "b", "c"] },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!("a")).is_ok());
    assert!(!schema.safe_parse(&json!("d")).success);
}

#[test]
fn import_array() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "array", "items": { "kind": "int" }, "minItems": 1 },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!([1, 2])).is_ok());
    assert!(!schema.safe_parse(&json!([])).success);
}

#[test]
fn import_tuple() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "tuple", "elements": [{ "kind": "string" }, { "kind": "int" }] },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!(["hello", 42])).is_ok());
}

#[test]
fn import_record() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "record", "values": { "kind": "string" } },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!({"a": "b"})).is_ok());
}

#[test]
fn import_union() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "union", "variants": [{ "kind": "string" }, { "kind": "int" }] },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!("hello")).is_ok());
    assert!(schema.parse(&json!(42)).is_ok());
}

#[test]
fn import_intersection() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "intersection", "allOf": [{ "kind": "number", "min": 0 }, { "kind": "number", "max": 100 }] },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!(50)).is_ok());
}

#[test]
fn import_optional() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "optional", "schema": { "kind": "string" } },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!("hello")).is_ok());
}

#[test]
fn import_nullable() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "nullable", "schema": { "kind": "string" } },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    assert!(schema.parse(&json!(null)).is_ok());
    assert!(schema.parse(&json!("hello")).is_ok());
}

#[test]
fn import_unknown_keys_strip() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": {
            "kind": "object",
            "properties": { "name": { "kind": "string" } },
            "required": ["name"],
            "unknownKeys": "strip"
        },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    let result = schema.parse(&json!({"name": "Alice", "extra": true}));
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), json!({"name": "Alice"}));
}

#[test]
fn import_unknown_keys_allow() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": {
            "kind": "object",
            "properties": { "name": { "kind": "string" } },
            "required": ["name"],
            "unknownKeys": "allow"
        },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    let result = schema.parse(&json!({"name": "Alice", "extra": true}));
    assert!(result.is_ok());
}

#[test]
fn roundtrip_export_import() {
    let original = object()
        .field("name", Box::new(string().min_length(1)))
        .field("age", Box::new(int().min(0.0)))
        .required(vec!["name", "age"]);

    let doc = export(&original, ExportMode::Portable, &HashMap::new()).unwrap();
    let json_str = interchange::exporter::document_to_json(&doc).unwrap();

    let (imported, _ctx) = import(&json_str).unwrap();
    assert!(imported
        .parse(&json!({"name": "Alice", "age": 30}))
        .is_ok());
    assert!(!imported.safe_parse(&json!({"name": "", "age": 30})).success);
}

#[test]
fn import_from_value_works() {
    let doc_value = json!({
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "string" },
        "definitions": {},
        "extensions": {}
    });
    let (schema, _ctx) = import_value(&doc_value).unwrap();
    assert!(schema.parse(&json!("hello")).is_ok());
}

#[test]
fn import_rejects_unsupported_kind() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": { "kind": "fantasy_type" },
        "definitions": {},
        "extensions": {}
    }"#;
    let result = import(json_str);
    assert!(result.is_err());
    assert!(result.unwrap_err().contains("Unsupported schema kind"));
}

#[test]
fn import_rejects_missing_kind() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": {},
        "definitions": {},
        "extensions": {}
    }"#;
    let result = import(json_str);
    assert!(result.is_err());
}

#[test]
fn import_default_with_object() {
    let json_str = r#"{
        "anyvaliVersion": "1.0",
        "schemaVersion": "1",
        "root": {
            "kind": "object",
            "properties": {
                "role": { "kind": "string", "default": "user" }
            },
            "required": [],
            "unknownKeys": "reject"
        },
        "definitions": {},
        "extensions": {}
    }"#;
    let (schema, _ctx) = import(json_str).unwrap();
    let result = schema.parse(&json!({}));
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), json!({"role": "user"}));
}
